### PR TITLE
DataloaderSegmentation == New dataloder

### PR DESCRIPTION
**Pull request by BlanchonMarc**
Addition of the dataloader as asked in issues. Inspired from piwise, simplified and with no other function than classes. Following Python2.XX and PEP8.

### CODE REVIEW

File: database/dataloaderSegmentation.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
+
```
Comment (glemaitre) [inline database/dataloaderSegmentation.py]: Remove this empty line

File: database/dataloaderSegmentation.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+'''
+Creating a Dataloader for segmentation taks
+'''
+
+import numpy as np
+import os
```
Comment (glemaitre) [inline database/dataloaderSegmentation.py]: standard library go first followed by third party library followed by local import

File: database/dataloaderSegmentation.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+
+class DatasetLoaderSegmentation(Dataset):
+    ''' Derived class to manage Segmentation Dataset Image Loading
+    '''
+
+    def __init__(self, root,
+                 images='images', imagext='png', imageconv='RGB',
+                 targets='targets', targetext='png', targetconv='P'):
+        # parameters copies
```
Comment (glemaitre) [inline database/dataloaderSegmentation.py]: this comment is not useful

File: database/dataloaderSegmentation.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+'''
+
+import numpy as np
+import os
+from PIL import Image
+from torch.utils.data import Dataset
+
+
+class DatasetLoaderSegmentation(Dataset):
+    ''' Derived class to manage Segmentation Dataset Image Loading
```
Comment (glemaitre) [inline database/dataloaderSegmentation.py]: This docstring is not informative enough.
As mentioned in one of the torchvision issue, I don't think that we should call it loader, because loaders are agnostic of the task (segmentation, classification. Probably for the moment, I would `ImageFolderSegmentation`. But I don't like the name. I would need to think about it.

Comment (glemaitre) [inline database/dataloaderSegmentation.py]: Derived class to manage Segmentation Dataset Image Loading -> We usually don't care to mention that it is derived. We see it from the code. This is more important to focus on what the class is doing itself

Comment (glemaitre) [reply database/dataloaderSegmentation.py]: You need to improve the docstring:


```python
class ImageFolderSegmentation(Dataset):
"""
    A generic data loader for image segmentation where the data are organised as:

    ....

    Parameters
    ----------
    root : str
        Root folder containing the segmentation data.
    images : str, defauit='images'
        ...

    Attributes
    ----------
    filenames : str
        The filenames of the data.
"""
```

So I just mention one parameter, one attribute to show you what they are.

File: database/dataloaderSegmentation.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    '''
+
+    def __init__(self, root,
```
Comment (glemaitre) [inline database/dataloaderSegmentation.py]: I don't think that we need root, images, targets, targetext.

We can simplify using wildcard and 2 parameters:
 
* `image_path` without any default. It can take a path as `'/some/path/*.png'`
* `label_path` without any default. It can take a path as `'/some/path/*.png'`

Comment (glemaitre) [reply database/dataloaderSegmentation.py]: label is better than target for segmentation I think

File: database/dataloaderSegmentation.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+                 targets='targets', targetext='png', targetconv='P'):
+        # parameters copies
+        self.images_root = os.path.join(root, images)
+        self.labels_root = os.path.join(root, targets)
+        self.ext = [imagext, targetext]
+
+        # Extract the filenames without extansions
+        self.filenames = [os.path.basename(os.path.splitext(fname)[0])
+                          for fname in os.listdir(self.labels_root)
+                          if any(filename.endswith(ext) for ext in self.ext)]
```
Comment (glemaitre) [inline database/dataloaderSegmentation.py]: not sure what is `filename` here

File: database/dataloaderSegmentation.py [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    def __init__(self, root,
+                 images='images', imagext='png', imageconv='RGB',
+                 targets='targets', targetext='png', targetconv='P'):
+        # parameters copies
+        self.images_root = os.path.join(root, images)
+        self.labels_root = os.path.join(root, targets)
+        self.ext = [imagext, targetext]
+
+        # Extract the filenames without extansions
+        self.filenames = [os.path.basename(os.path.splitext(fname)[0])
```
Comment (glemaitre) [inline database/dataloaderSegmentation.py]: you can use directly glob here

``` python
self.image_filenames = sorted(glob.glob(self.image_path))
self.label_filenames = sorted(glob.glob(self.label_path))

def _get_filename(path):
    return os.path.basename(os.path.splitext(path)[0]) 

if not all([_get_filename(if) == _get_filename(lf)
            for if, lf in zip(image_filenames, label_filenames)]
    raise ValueError('Some files are different.')
```

File: database/dataloaderSegmentation.py [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+                          for fname in os.listdir(self.labels_root)
+                          if any(filename.endswith(ext) for ext in self.ext)]
+
+        # Alphabetical Order
+        self.filenames.sort()
```
Comment (glemaitre) [inline database/dataloaderSegmentation.py]: useless since we sorted before

File: database/dataloaderSegmentation.py [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+
+    def __getitem__(self, index):
+        '''Convert the images and return to be passed in dataloader
+
+        Dataloader will recieve the images and transform
+        It will also allow the paralellization / optimization
+        '''
+
+        # Images
+        with open(os.path.join(self.images_root, self.filenames[index] \
```
Comment (glemaitre) [inline database/dataloaderSegmentation.py]: avoid to use `\`

File: database/dataloaderSegmentation.py [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+from PIL import Image
+from torch.utils.data import Dataset
+
+
+class DatasetLoaderSegmentation(Dataset):
+    ''' Derived class to manage Segmentation Dataset Image Loading
+    '''
+
+    def __init__(self, root,
+                 images='images', imagext='png', imageconv='RGB',
```
Comment (glemaitre) [inline database/dataloaderSegmentation.py]: Do we need to have the conversion here. Why is it not enough to be hard-coded.

File: database/dataloaderSegmentation.py (hunk 9 above)
Comment (glemaitre) [inline database/dataloaderSegmentation.py]: so now it should be simplified as:

```python
with open(image_filenames[index], 'rb') as f:
    image = Image.open(f).convert('RGB')
```

File: database/dataloaderSegmentation.py [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+        It will also allow the paralellization / optimization
+        '''
+
+        # Images
+        with open(os.path.join(self.images_root, self.filenames[index] \
+                               + self.ext[0]), 'rb') as f:
+            image = Image.open(f).convert(imageconv)
+
+        # Targets
+        with open(os.path.join(self.labels_root, self.filenames[index] \
```
Comment (glemaitre) [inline database/dataloaderSegmentation.py]:
```python
with open(label_filenames[index], 'rb') as f:
    label = Image.open(f).convert('P')
```

File: database/dataloaderSegmentation.py [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+        Dataloader will recieve the images and transform
+        It will also allow the paralellization / optimization
+        '''
+
+        # Images
+        with open(os.path.join(self.images_root, self.filenames[index] \
+                               + self.ext[0]), 'rb') as f:
+            image = Image.open(f).convert(imageconv)
+
+        # Targets
```
Comment (glemaitre) [inline database/dataloaderSegmentation.py]: no need for comment

File: database/dataloaderSegmentation.py [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+        self.filenames.sort()
+
+    def __getitem__(self, index):
+        '''Convert the images and return to be passed in dataloader
+
+        Dataloader will recieve the images and transform
+        It will also allow the paralellization / optimization
+        '''
+
+        # Images
```
Comment (glemaitre) [inline database/dataloaderSegmentation.py]: no need for comment

File: database/dataloaderSegmentation.py [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+        with open(os.path.join(self.images_root, self.filenames[index] \
+                               + self.ext[0]), 'rb') as f:
+            image = Image.open(f).convert(imageconv)
+
+        # Targets
+        with open(os.path.join(self.labels_root, self.filenames[index] \
+                               + self.ext[1]), 'rb') as f:
+            target = Image.open(f).convert(targetconv)
+
+        return image, target
```
Comment (glemaitre) [inline database/dataloaderSegmentation.py]: target ->label

File: database/dataloaderSegmentation.py [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+        # Extract the filenames without extansions
+        self.filenames = [os.path.basename(os.path.splitext(fname)[0])
+                          for fname in os.listdir(self.labels_root)
+                          if any(filename.endswith(ext) for ext in self.ext)]
+
+        # Alphabetical Order
+        self.filenames.sort()
+
+    def __getitem__(self, index):
+        '''Convert the images and return to be passed in dataloader
```
Comment (glemaitre) [inline database/dataloaderSegmentation.py]: I think that `Get an image and a label will be enough`

File: database/dataloaderSegmentation.py [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+
+        # Targets
+        with open(os.path.join(self.labels_root, self.filenames[index] \
+                               + self.ext[1]), 'rb') as f:
+            target = Image.open(f).convert(targetconv)
+
+        return image, target
+
+    def __len__(self):
+        return len(self.filenames)
```
Comment (glemaitre) [inline database/dataloaderSegmentation.py]: `len(self.image_filenames)`

File: database/dataloaderSegmentation.py (hunk 7 above)
Comment (glemaitre) [reply database/dataloaderSegmentation.py]: Improve the error message, it is just a draft.

**Review by glemaitre — COMMENTED** [review]
Couple of comments. I think that you are missing the transform on the image and the label images

File: database/dataloaderSegmentation.py (hunk 10 above)
Comment (BlanchonMarc) [reply database/dataloaderSegmentation.py]: Addition of a variable to contain the conversions

**Comment by glemaitre** [comment]
As a general comment, a PR should be specific to a change. So programming `ImageFolderSegmentation` should be a PR and the change in the loader making python 2 compatible should be in another PR. It allows to simplify the review process.
